In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [2]:
df = pd.read_csv('data_eda.csv')

In [3]:
# choose relavent columns
df.columns
df_model = df[['Rating','Size', 'Type of ownership', 'Industry', 'Sector', 
               'Revenue','Salary_type','avg_salary','job_state', 'same_state', 
               'company_age','python_JD','rstudio_JD', 'spark_JD', 'aws_JD',
               'machinelearning_JD','datascience_JD', 'data_simp', 'seniority',
               'job_desc_len']]


In [4]:
# get dummy data
df_dummy = pd.get_dummies(df_model)

In [5]:
# train test split
from sklearn.model_selection import train_test_split

X = df_dummy.drop('avg_salary', axis=1)
Y = df_dummy.avg_salary.values

x_train, x_test, y_train, y_test = train_test_split(X,Y,test_size=0.20,random_state=1)

In [6]:
# mutiple linear regression
    # we are going to do the linear regression in statsmodels and will use their package in sklearn

import statsmodels.api as sm

X_sm = X = sm.add_constant(X)
model = sm.OLS(Y,X_sm)
model.fit().summary() # it will give the complte statistical modeling calculation summary

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.934
Model:                            OLS   Adj. R-squared:                  0.917
Method:                 Least Squares   F-statistic:                     53.26
Date:                Mon, 25 May 2020   Prob (F-statistic):          3.01e-177
Time:                        20:18:49   Log-Likelihood:                -1671.6
No. Observations:                 489   AIC:                             3551.
Df Residuals:                     385   BIC:                             3987.
Df Model:                         103                                         
Covariance Type:            nonrobust                                         
======================================================================================================================
                                                         coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------------
const                                                 18.7530      5.811      3.227      0.001       7.328      30.178
Rating                                                -0.8912      1.463     -0.609      0.543      -3.767       1.985
same_state                                            12.6456      2.491      5.076      0.000       7.747      17.544
company_age                                            0.0350      0.033      1.054      0.293      -0.030       0.100
python_JD                                             -6.2784      2.357     -2.664      0.008     -10.912      -1.644
rstudio_JD                                            -6.4455      6.107     -1.055      0.292     -18.454       5.563
spark_JD                                              -5.2370      2.505     -2.091      0.037     -10.161      -0.313
aws_JD                                                 1.8640      2.572      0.725      0.469      -3.194       6.921
machinelearning_JD                                    16.2020      2.671      6.065      0.000      10.950      21.454
datascience_JD                                        -5.6390      2.508     -2.249      0.025     -10.570      -0.709
job_desc_len                                           0.0003      0.001      0.330      0.741      -0.001       0.002
Size_1 to 50 employees                                 7.5573      4.599      1.643      0.101      -1.484      16.599
Size_10000+ employees                                -12.1630      4.239     -2.869      0.004     -20.498      -3.828
Size_1001 to 5000 employees                           -7.4870      2.780     -2.693      0.007     -12.954      -2.020
Size_201 to 500 employees                             -0.6778      3.154     -0.215      0.830      -6.879       5.523
Size_5001 to 10000 employees                          -1.0179      4.313     -0.236      0.814      -9.498       7.462
Size_501 to 1000 employees                             1.5773      3.164      0.499      0.618      -4.643       7.797
Size_51 to 200 employees                               9.6754      3.523      2.746      0.006       2.749      16.602
Size_Unknown                                          21.2888      9.684      2.198      0.029       2.249      40.328
Type of ownership_-1                                   6.0275      5.430      1.110      0.268      -4.649      16.704
Type of ownership_Company - Private                   -3.4202      3.318     -1.031      0.303      -9.943       3.103
Type of ownership_Company - Public                     6.4723      3.716      1.742      0.082      -0.834      13.778
Type of ownership_Government                           0.9643     11.027      0.087      0.930     -20.716  

this above output is all statistical output of the data, it provide all statistical terms, there are very useful so lets understand it, but keep in mind
Data science is more focused on predictions and having models with predict on real words data, it is not as much concerned about correct specification of statistical problems.

**Dep. Variable:** this is target variable for which we are predicting the models, generally it is on Y axis  
**Model:** this is model name we used in this regression problem.  
**Method:** Method which was used to achieving the coefficient estimates, Least squares is the standard method (method used to fit the data into model)  
**No. Observations:** number of  observarions  
**Df Residuals:** the degree of freedom(DF) of the residuals, calculated by thaking the numbr of observations less than the number of parameters   
**Df Model:** the number of estimated parameters in the model, (Degree of freedom of the model)

**No. Observations = Df Residuals + Df Model + 1 (constant always left)**


**R-squared:** this is the coefficient of derermination, it measures the goodness of fit. (it basically tells the how much your data is fitted in the model)  

$$R^2 = 1 - \frac{SS_{res}}{SS_{tot}} $$  

**Adj. R-squared:** this is the adjusted R-squared. it is coefficient of determination adjusted by sample size and the number of parameter used. Adj R-squared is generally used in multiple regression, not in linear regression
$$ \bar{R}^2 = 1 - (1-R^2)\frac{n-1}{n-p-1} $$

> p = the total number of explanatory variables  
> n = the sample size

**F-statistic:** A measure that tells you if your model is differnt form simple average  

**Prob (F-statistic):** this measures the significance of your F-statistic, also called p-value of F-statistic, in statistics p-value equal or lower than 0.05 is considered significant.  

**AIC:** This is the Akike information criterion, it evaluates the model based on model complexity and number of observation, lower is better.  

**BIC:** This is Bayesian information criterion, similar to AIC, except it pushesh the model with more parameters

### Parameters  Estimated and the Associated Statistical Tests
**coef:** : The estimated coefficient test, note that this is just a point estiamtes  

**std err:** standard error if the estimate of the coefficient, Another term of standard deviation  

**t:** the t-statistic score  

**p>|t|:** The p-value, A measure of the probability that the coefficient is different from zreo. p-value greater then 0.05 if insignificant.  

**[95.0% Conf. Interval]:** The 95% confidience interval of the coefficient, Shown here as [0.025, 0.975]